File: rag-langchain-python/ask_contextual_query.py
# Based on example at
# https://python.langchain.com/docs/use_cases/question_answering/quickstart .

import os
import sys
from dotenv import load_dotenv
from langchain import hub
from langchain_openai import AzureChatOpenAI
from langchain.schema import StrOutputParser
from langchain.schema.runnable import RunnablePassthrough
from marklogic import Client
from contextual_query_retriever import ContextualQueryRetriever


def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)


question = sys.argv[1]

retriever = ContextualQueryRetriever.create(
    Client("http://localhost:8003", digest=("ai-examples-user", "password"))
)

load_dotenv()

prompt = hub.pull("rlm/rag-prompt")
# Note that the Azure OpenAI API key, the Azure OpenAI Endpoint, and the OpenAI API
# Version, are all read from the environment automatically.
llm = AzureChatOpenAI(
    model_name=os.getenv("AZURE_LLM_DEPLOYMENT_NAME"),
    azure_deployment=os.getenv("AZURE_LLM_DEPLOYMENT_NAME"),
    temperature=0,
    max_tokens=None,
    timeout=None,
)

contextual_query = {
    "query": {
        "queries": [
            {
                "value-query": {
                    "json-property": "type",
                    "text": "public intoxication"
                }
            }
        ]
    }
}
chat_context = {"question": question, "contextual_query": contextual_query}

rag_chain = (
    {
        "context": retriever | format_docs,
        "question": RunnablePassthrough().pick("question"),
    }
    | prompt
    | llm
    | StrOutputParser()
)

print(rag_chain.invoke(input=chat_context))

File: rag-langchain-js/askContextualQuery.js
import { config } from 'dotenv';
import { AzureChatOpenAI } from "@langchain/openai";

import {StringOutputParser} from "@langchain/core/output_parsers";
import {createStuffDocumentsChain} from "langchain/chains/combine_documents";
import * as hub from "langchain/hub";
import {createDatabaseClient} from "marklogic";
import {ContextualQueryRetriever} from "./contextualQueryRetriever.js";

config({path: "../.env"});
const marklogicClient = createDatabaseClient({
  host:     'localhost',
  port:     '8003',
  user:     'ai-examples-user',
  password: 'password',
  authType: 'DIGEST'
});
const contextualQueryRetriever = new ContextualQueryRetriever({marklogicClient});

const llm = new AzureChatOpenAI({ });
const prompt = await hub.pull("rlm/rag-prompt");

let question = "What disturbances has Jane Doe caused?";
if (process.argv.length > 2) {
  question = process.argv[2];
}
console.log(`Question: ${question}`);

const contextualQuery = {
  "query": {
    "queries": [
      {
        "value-query": {
          "json-property": "type",
          "text": "public intoxication"
        }
      }
    ]
  }
}

const ragChain = await createStuffDocumentsChain({
  llm,
  prompt,
  outputParser: new StringOutputParser()
});
const chainResponse = await ragChain.invoke({
  question: question,
  context: await contextualQueryRetriever.invoke({"question": question, "contextualQuery": contextualQuery}),
});
console.log("\n\nChatbot VectorQuery chainResponse: \n" + JSON.stringify(chainResponse));
